Source/Core/DolphinQt/Settings.cpp

```diff
@@ -40,8 +42,17 @@ Settings::Settings()
     QueueOnObject(this, [this, new_state] { emit EmulationStateChanged(new_state); });
   });
 
-  Config::AddConfigChangedCallback(
-      [this] { QueueOnObject(this, [this] { emit ConfigChanged(); }); });
+  Config::AddConfigChangedCallback([this] {
+    static std::atomic<bool> do_once{true};
+    if (do_once)
+    {
+      do_once = false;
```

Comment:
This doesn't seem very race-free to me. I believe you should at least make the following change, though I'm not sure if it's sufficient or if something also needs to be changed at the line where you're assigning `false`:

```suggestion
    if (do_once.exchange(false))
    {
```

Reply:
It should be, as long as do_once is set to true before emitting the OnConfigChanged() I don't see any problem. Even that exchange wasn't necessary for safety, but it's an optimization if multiple threads changed settings at the same time.

Reply:
I believe you're correct in that the only harmful effect of this race is that we will emit ConfigChanged more times than needed.

Reply:
When "Config" invokes the ConfigChanged() callbacks, the settings have already changed, so even if the host/Qt thread callstack was within that queued lambda for a previous config change, the setting change will either be broadcast immediately within the current Qt OnConfigChanged() or queued for later.
The only race condition possible is that the optimization would fail to apply and OnConfigChanged() is called more than needed.